synthesizer/process/src/stack/mod.rs

```diff
@@ -215,54 +224,164 @@ pub struct Stack<N: Network> {
+        // Check that the program is well-formed.
+        check_program_is_well_formed(program)?;
+
+        // If the program exists in the process, check that the new program is valid.
         if let Ok(existing_stack) = process.get_stack(program_id) {
             // Ensure the program is not `credits.aleo`.
             ensure!(program_id != &ProgramID::from_str("credits.aleo")?, "Cannot re-initialize the 'credits.aleo'.");
-            // Ensure that the new program matches the existing program.
+            // Get the existing program.
+            let existing_program = existing_stack.program();
+            // If the existing program does not have a constructor, check that the new program matches the existing program.
+            // Otherwise, ensure that the upgrade is valid.
+            match existing_program.contains_constructor() {
+                false => ensure!(
+                    existing_stack.program() == program,
+                    "Program '{program_id}' already exists with different contents."
+                ),
+                true => Self::check_upgrade_is_valid(existing_program, program)?,
+            }
+        }
+
+        // Return the stack.
+        Stack::initialize(process, program)
+    }
+
+    /// Partially initializes a new stack, given the process and the program, without checking for validity.
+    /// Note. This method should **NOT** be used by the on-chain VM to add new program, use `Stack::new` instead.
+    pub fn new_raw(process: &Process<N>, program: &Program<N>, edition: u16) -> Result<Self> {
+        // Check that the program is well-formed.
+        check_program_is_well_formed(program)?;
+        // Return the stack.
+        Stack::create_raw(process, program, edition)
+    }
+
+    /// Checks and initializes the register state of the stack, even if it has already been initialized.
+    pub fn check_and_initialize(&self, process: &Process<N>) -> Result<()> {
+        // Clear the existing constructor, closure, and function types.
+        self.constructor_types.write().take();
+        self.register_types.write().clear();
+        self.finalize_types.write().clear();
+
+        // Add all the imports into the stack.
+        for import in self.program.imports().keys() {
+            // Ensure that the program does not import itself.
+            ensure!(import != self.program.id(), "Program cannot import itself");
+            // Ensure the program imports all exist in the process already.
+            if !process.contains_program(import) {
+                bail!("Cannot add program '{}' because its import '{import}' must be added first", self.program.id())
+            }
+        }
+
+        // Add the constructor to the stack if it exists.
+        if let Some(constructor) = self.program.constructor() {
+            // Ensure that the constructor is not already added.
+            ensure!(self.constructor_types.read().is_none(), "Constructor already exists");
```

Comment:
this _should_ be redundant due to line 276; otherwise, the write guard should be kept for longer